### PR TITLE
Improve local (native) build and Cross compile for Windows

### PR DESCRIPTION
**Pull request by networkfusion**
This works on a native cross compile for windows and as a side effect also improves the ability to build the toolchain locally (via ubuntu on wsl2 in my instance).

I am still working out the kinks (since it takes so long to build) but it also contains improved comments for the native build.

### CODE REVIEW

**Comment by networkfusion** [comment]
Current progress:

Now builds on WSL2 with the following caveats:
* linux build must have been built and installed first, otherwise it fails with an error related to: 
```mips64-elf-gcc -dumpspecs > tmp-specs
/bin/bash: mips64-elf-gcc: command not found
make[1]: *** [Makefile:2089: specs] Error 127
```
Although that "could" be a path issue OR not able to use `mips64-elf-gcc.exe`
* Requires compiling and specifying gmp, mpfr and mpc otherwise they cannot be found (probably a path problem as only uses the libs?!)
* Because of having to compile `gmp`, on WSL there are issues (solved using https://github.com/crosstool-ng/crosstool-ng/issues/1328 as reference).
* make uses 4.2.1 due to a `dirent.h` regression in 4.3 (`conflicting types for ‘fcntl’`)

**Comment by networkfusion** [comment]
@rasky I have now got the FP libs working under the GCC source tree...
Current issues:
* linux build must have been built and installed first, otherwise it fails with an error related to: 
```mips64-elf-gcc -dumpspecs > tmp-specs
/bin/bash: mips64-elf-gcc: command not found
make[1]: *** [Makefile:2089: specs] Error 127
```
Although that "could" be a path issue OR not able to use `mips64-elf-gcc.exe`
* If the above cannot be solved, work towards using a cache package of the linux job has been attempted (although I might need help there).
* I tried to use GCC 11.2 but ran into errors relating to rust! so for the moment still uses GCC 10.2 and binutils 2.36.1 as known working.
* make uses 4.2.1 due to a `dirent.h` regression in 4.3 (`conflicting types for ‘fcntl’`)

**Comment by rasky** [comment]
@networkfusion great news!

Can you please show me a log where I can see this error: 
```
/bin/bash: mips64-elf-gcc: command not found
make[1]: *** [Makefile:2089: specs] Error 127
```

I want to investigate

**Comment by networkfusion** [comment]
> 
> 
> @networkfusion great news!
> 
> Can you please show me a log where I can see this error:
> 
> ```
> make[1]: *** [Makefile:2089: specs] Error 127```
> 
> I want to investigate
> ```

https://github.com/networkfusion/libdragon/runs/4142506129?check_suite_focus=true was the last try, although there were a number of attempts before reverting to using a preinstalled linux lib. Please check my action history, although not great commit naming, it does tell a story 😁

**Comment by rasky** [comment]
I found the mention of this problem in the GCC docs:
https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html

```
If build and host are different, you must have already built and installed a cross compiler that will be used to build the target libraries (if you configured with --target=foo-bar, this compiler will be called foo-bar-gcc).
```

which makes sense. The process must compile target libraries such as libstdc++, and those require a linux->mips compiler to be available, as the win->mips compiler that was just built cannot be run on the build machine.

I'll review the PR.

File: tools/build-toolchain.sh [hunk 1]
```diff
@@ -1,34 +1,71 @@
 #! /bin/bash
-# N64 MIPS GCC toolchain build/install script for Unix distributions
+# N64 MIPS GCC toolchain build/install script.
```
Comment (rasky) [inline tools/build-toolchain.sh]: Please do not do this kind of unrelated edits to comments. Let's try to keep the changes to the *bare minimum*. If you want to improve documentation, do it in a different PR.

File: tools/build-toolchain.sh [hunk 2]
```diff
@@ -1,34 +1,71 @@
...
 # (c) 2012-2021 Shaun Taylor and libDragon Contributors.
 # See the root folder for license information.
 
 
+# This script may prompt for a password if it is not called with elevated privileges.
+# Preferably it is fixed at the start!
+#TODO: ensure elevated privileges! e.g.
+#echo "$(whoami)"
+#[ "$UID" -eq 0 ] || exec sudo "$0" "$@"
```
Comment (rasky) [inline tools/build-toolchain.sh]: Please remove also this, or move it to a different PR.

File: tools/build-toolchain.sh [hunk 3]
```diff
@@ -1,34 +1,71 @@
...
+# sudo apt-get update && sudo apt-get upgrade
+# sudo apt-get install -yq wget bzip2 gcc g++ make file libmpfr-dev libmpc-dev zlib1g-dev texinfo git gcc-multilib
 
 # Exit on error
 set -e
 
-# Set N64_INST before calling the script to change the default installation directory path
-INSTALL_PATH="${N64_INST:-/usr/local}"
+# Check for cross compile script flag
+if [ "$1" == "-xcw" ]; then # Windows cross compile flag is specified as a parameter.
```
Comment (rasky) [inline tools/build-toolchain.sh]: Can we please change this special command line to simply be `--host x86_64-w64-mingw32`? This way we can reuse for different architectures, if needed.

File: tools/build-toolchain.sh [hunk 4]
```diff
@@ -1,34 +1,71 @@
...
+  # This (probably) requires the toolchain to have already built and installed on the native (linux) system first!
+  # This (may) also require the following (extra) package dependencies:
+  # sudo apt-get install -yq mingw-w64 libgmp-dev bison libz-mingw-w64-dev autoconf
+
+  echo "Cross compiling for Windows x86_64"
+  # Use the current-directory/win_binaries for the install path, as this is not for linux!
+  rm -rf win_binaries
+  mkdir win_binaries
+  CURRENT_PATH="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+  INSTALL_PATH="$CURRENT_PATH/win_binaries"
```
Comment (rasky) [inline tools/build-toolchain.sh]: Can you please explain why this is needed?

File: tools/build-toolchain.sh [hunk 5]
```diff
@@ -1,34 +1,71 @@
...
+  # Dependency source libs (Versions)
+  # This will have to build Make and download FP libs
+  GMP_V=6.2.0
+  MPC_V=1.2.1
+  MPFR_V=4.1.0
+  MAKE_V=4.2.1
+  # These "should" be the same as linux, but may be out of sync (as need to ensure working natively first).
+  BINUTILS_V=2.36.1
+  GCC_V=10.2.0
+  NEWLIB_V=4.1.0
```
Comment (rasky) [inline tools/build-toolchain.sh]: Please remove this. We need to keep this in sync between Linux and Windows, we don't want to have two different versions of binutils or GCC or newlib.

File: tools/build-toolchain.sh [hunk 6]
```diff
@@ -1,34 +1,71 @@
...
+  INSTALL_PATH="$CURRENT_PATH/win_binaries"
+
+  # This will require the extra flags (under certain libs)
+  CROSS_COMPILE_FLAGS="--build=x86_64-linux-gnu --host=x86_64-w64-mingw32"
+
+  # Dependency source libs (Versions)
+  # This will have to build Make and download FP libs
+  GMP_V=6.2.0
+  MPC_V=1.2.1
+  MPFR_V=4.1.0
```
Comment (rasky) [inline tools/build-toolchain.sh]: Please define these symbols as empty on Linux, such as `GMP_V=`, instead of keeping the variable as undefined there.

File: tools/build-toolchain.sh [hunk 7]
```diff
@@ -50,23 +87,42 @@ download () {
 test -f "binutils-$BINUTILS_V.tar.gz" || download "https://ftp.gnu.org/gnu/binutils/binutils-$BINUTILS_V.tar.gz"
 test -f "gcc-$GCC_V.tar.gz"           || download "https://ftp.gnu.org/gnu/gcc/gcc-$GCC_V/gcc-$GCC_V.tar.gz"
 test -f "newlib-$NEWLIB_V.tar.gz"     || download "https://sourceware.org/pub/newlib/newlib-$NEWLIB_V.tar.gz"
+if [ "$CROSS_COMPILE_FLAGS" != "" ]; then
+  test -f "make-$MAKE_V.tar.gz"         || download "https://ftp.gnu.org/gnu/make/make-$MAKE_V.tar.gz"
+  test -f "gmp-$GMP_V.tar.xz"           || download "https://ftp.gnu.org/gnu/gmp/gmp-$GMP_V.tar.xz"
+  test -f "mpc-$MPC_V.tar.gz"           || download "https://ftp.gnu.org/gnu/mpc/mpc-$MPC_V.tar.gz"
+  test -f "mpfr-$MPFR_V.tar.gz"         || download "https://ftp.gnu.org/gnu/mpfr/mpfr-$MPFR_V.tar.gz"
+fi
```
Comment (rasky) [inline tools/build-toolchain.sh:124]: Do not use CROSS_COMPILE_FLAGS for this. Do a test for each package, checking that the variable is defined and not empty. Such as `If ["$GMP_V" != ""]`

**Comment by networkfusion** [comment]
As a note, I have partly used the https://github.com/DragonMinded/libdragon/commit/799b60e9a997658f911d5a3d5de4675e22da006d commit against this PR. It will be helpful for reference, since I wanted to ensure that it did not effect the linux build (since it "might" still effect the windows build (still waiting on WSL builds)...

File: tools/build-toolchain.sh [hunk 8]
```diff
@@ -50,23 +87,42 @@ download () {
 test -f "binutils-$BINUTILS_V.tar.gz" || download "https://ftp.gnu.org/gnu/binutils/binutils-$BINUTILS_V.tar.gz"
 test -f "gcc-$GCC_V.tar.gz"           || download "https://ftp.gnu.org/gnu/gcc/gcc-$GCC_V/gcc-$GCC_V.tar.gz"
 test -f "newlib-$NEWLIB_V.tar.gz"     || download "https://sourceware.org/pub/newlib/newlib-$NEWLIB_V.tar.gz"
+if [ "$CROSS_COMPILE_FLAGS" != "" ]; then
+  test -f "make-$MAKE_V.tar.gz"         || download "https://ftp.gnu.org/gnu/make/make-$MAKE_V.tar.gz"
+  test -f "gmp-$GMP_V.tar.xz"           || download "https://ftp.gnu.org/gnu/gmp/gmp-$GMP_V.tar.xz"
```
Comment (rasky) [inline tools/build-toolchain.sh]: Please use always the `.tar.gz` version if possible. We do not want to depend on `xz` if not required.

File: tools/build-toolchain.sh [hunk 9]
```diff
@@ -50,23 +87,42 @@ download () {
...
+  test -d "make-$MAKE_V"         || tar -xzf "make-$MAKE_V.tar.gz"
+  test -d "gmp-$GMP_V"           || tar -xf "gmp-$GMP_V.tar.xz"
+  test -d "mpc-$MPC_V"           || tar -xzf "mpc-$MPC_V.tar.gz"
+  test -d "mpfr-$MPFR_V"         || tar -xzf "mpfr-$MPFR_V.tar.gz"
+
+  # Copies the FP libs into GCC sources so they are compiled as part of it
+  cp -R "gmp-$GMP_V" "gcc-$GCC_V"/gmp
+  cp -R "mpc-$MPC_V" "gcc-$GCC_V"/mpc
+  cp -R "mpfr-$MPFR_V" "gcc-$GCC_V"/mpfr
+fi
```
Comment (rasky) [inline tools/build-toolchain.sh:150]: Also here please do one check for package. Maybe collapse with the download check, so that you don't repeat all the checks here. Just one check for download + unpack.

Also don't do the copy: just create a symlink with `ln -s`.

File: tools/build-toolchain.sh [hunk 10]
```diff
@@ -87,30 +143,34 @@ cd gcc_compile
...
   --disable-nls \
   --disable-werror \
-  --with-system-zlib
+  --with-system-zlib \
+  $CROSS_COMPILE_FLAGS
 make all-gcc -j "$JOBS"
 make all-target-libgcc -j "$JOBS"
 make install-gcc || sudo make install-gcc || su -c "make install-gcc"
 make install-target-libgcc || sudo make install-target-libgcc || su -c "make install-target-libgcc"
+echo "Finished Compiling GCC-$GCC_V for MIPS N64 (pass 1) outside of the source tree"
```
Comment (rasky) [inline tools/build-toolchain.sh]: I think the idea of having 2 GCC passes was to support exactly this scenario. So let's try something different.

The first two steps (binutils+GCC) should build a compiler for the **build** platform, so for Linux. Don't add your `$CROSS_COMPILE_FLAGS` here. This is what will build and install the linux->mips compiler that you need later to avoid the error. Then you build newlib. That will be done with the linux->mips compiler, so specifying `$CROSS_COMPILE_FLAGS` should not be required as well.

Now, before the final pass (GCC step 2), you probably need to add a new pass to build binutils *again*, but this time with $CROSS_COMPILE_FLAGS. This will build the windows->mips binutils which are required. And then, the final pass with GCC with $CROSS_COMPILE_FLAGS, that should succeed because the step 1 has installed the pre-required linux->mips compiler.

Hope this makes sense.

To try this, please also modify the CI to have a separate job that *only* does the Windows build.

File: tools/build-toolchain.sh [hunk 11]
```diff
@@ -124,6 +184,29 @@ CFLAGS_FOR_TARGET="-G0 -O2" CXXFLAGS_FOR_TARGET="-G0 -O2" ../"gcc-$GCC_V"/config
...
+  ./configure \
+    --prefix="$INSTALL_PATH" \
+    --disable-largefile \
+    --disable-nls \
+    --disable-rpath \
+    $CROSS_COMPILE_FLAGS
 make -j "$JOBS"
 make install || sudo make install || su -c "make install"
+echo "Finished Compiling make-$MAKE_V"
+fi
```
Comment (rasky) [inline tools/build-toolchain.sh:287]: I don't think we should build make here, also because it's not enough: n64.mk also requires bash and other tools. Maybe we can just require a shell like msys to be installed on Windows? msys already contains make and other required tools.

**Review by rasky — CHANGES_REQUESTED** [review]
As a general note, please squash all your commits and rebase over the current trunk. Let's start with a clean history.

I will push an updated trunk so that you can sync.

File: tools/build-toolchain.sh (hunk 8 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: Unfortunately there is no gz package for this lib 😒 
image
ref: https://ftp.gnu.org/gnu/gmp/

File: tools/build-toolchain.sh (hunk 1 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: Surely it is related since the script can now cross compile for other platforms?

Comment (rasky) [reply tools/build-toolchain.sh]: Well, I was reading it meaning that this script is meant to run on Unix distributions. Nevermind though, we can keep this change.

File: tools/build-toolchain.sh (hunk 4 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: This is because we want to use the `build-toolchain.sh`'s current directory as a basis for the install location for packaging and it should definitely not be installing any of the output to the current N64INST location... However it might be better to handle it in the github workflow script (although it will be harder to handle that for users who want to handle it). Although, it should be possible to combine them now, as they were originally split to handle MinGW libs.

File: tools/build-toolchain.sh (hunk 5 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: I still need to ensure this works under all environments. I had errors updating using binutils `2.37` when cross compiling.

File: tools/build-toolchain.sh (hunk 10 above)
Comment (rasky) [reply tools/build-toolchain.sh]: So to better clarify. Assume that you have defined two variables: `BUILD=x86_64-linux-gnu` and `HOST` that can be either `x86_64-linux-gnu` or `x86_64-w64-mingw32`. 

I think the workflow should be (pseudo-code)

```
if BUILD != HOST {
     Build binutils BUILD->TARGET
     Install binutils
     Build GCC BUILD->TARGET (no C++, only all-gcc/all-libgcc)  -- step1
     Install GCC
     Unpack/symlink GMP, MPFR, etc. -- preparation for next step where they are required only if BUILD != HOST
}

Build binutils BUILD/HOST/TARGET
if BUILD==HOST: Install binutils
Build GCC BUILD/HOST/TARGET with C,C++  -- step 2
if BUILD==HOST: Install GCC
Build newlib for TARGET
```

So this is actually exactly what I wrote in the previous comment, but it's just that checks are differents. You want an initial phase which is run only when BUILD != HOST which builds binutils+GCC as basis, and then the main phase which is always run that builds binutils+GCC+newlib for the BUILD/HOST/TARGET triplet.

If you structure like this, you can probably drop the custom command line and simply having the CI redefine the HOST environment. Something like:

```
BUILD=x86_64-linux-gnu
HOST=${HOST:-$BUILD}
TARGET=mips64-elf
```

and then the CI should be able to do `HOST=x86_64-w64-mingw build-toolchain.sh`

**Comment by networkfusion** [comment]
@rasky , I am taking ever larger steps towards your suggestions (I have tried a rebase and utimately faild, so will create a new PR)... But working on your Foreign BinUtils and GCC Pass 2 assumption the latest change fails with 
image

So would appreciate any suggestions there...

File: tools/build-toolchain.sh [hunk 12]
```diff
@@ -92,25 +159,64 @@ make all-gcc -j "$JOBS"
...
+make install || sudo env PATH="$PATH" make install || su -c "env PATH=\"$PATH\" make install" # Perhaps use `checkinstall` instead?!
+make distclean # Ensure we can build it again
+echo "Finished Compiling newlib-$NEWLIB_V"
+
+if [ "$BUILD" != "$HOST" ]; then
+  INSTALL_PATH="${FOREIGN_INSTALL_PATH}"
+fi
+
+if [ "$BUILD" != "$HOST" ]; then
+  echo "Compiling newlib-$NEWLIB_V for foreign host"
```
Comment (rasky) [inline tools/build-toolchain.sh]: Newlib is a target library. Why should we build it twice? It should be sufficient to build it once with either the first compiler

File: tools/build-toolchain.sh [hunk 13]
```diff
@@ -92,25 +159,64 @@ make all-gcc -j "$JOBS"
...
+  ./configure \
+    --prefix="$INSTALL_PATH" \
+    --target=mips64-elf \
+    --with-cpu=mips64vr4300 \
+     --disable-werror \
+    $BUILD \
+    $HOST
+  make -j "$JOBS"
+  make install || sudo make install || su -c "make install"
+  make distclean # Ensure we can build it again
```
Comment (rasky) [inline tools/build-toolchain.sh]: Why are you adding these distclean?

File: tools/build-toolchain.sh (hunk 12 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: It was not in the output (binary folder) expected when I built on WSL. I probably still have this wrong, but working on it...

File: tools/build-toolchain.sh (hunk 13 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: This is necessary, otherwise the binutils lib fails when trying to build for a different host. Although if this is doing the wrong thing, it would be necessary to extract twice so the build is targeted.

File: tools/build-toolchain.sh (hunk 11 above)
Comment (networkfusion) [reply tools/build-toolchain.sh:287]: My argument here would be that if you have to use MSYS (or an equivalent environment), it would totally negate the need for a native windows native toolchain. That said, this script does not (yet) build the dragon lib (libdragon), and older projects never had the need for Unix only utilities (like `dd`). I would suggest that `makefile` should be available, but caveats are added... OR the makefile build is just added in the CI script.

File: tools/build-toolchain.sh (hunk 9 above)
Comment (networkfusion) [reply tools/build-toolchain.sh:150]: My attempts so far against symbolic links have been unsuccessful, however I am attempting again against WSL using `ln -s -T "mpfr-$MPFR_V"/ "gcc-$GCC_V"/mpfr`

File: tools/build-toolchain.sh (hunk 13 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: Also, it (for all other make installs) it ensures that from a local target, the libs will successfully compile again without destroying the folder.

File: tools/build-toolchain.sh (hunk 11 above)
Comment (rasky) [reply tools/build-toolchain.sh:287]: The point is that n64.mk requires at least bash if, rm, mkdir and might grow even more. I don't think we can make it work without Linux tools. Installing msys is a fast way for the user to have those tools available, as far as I know. If you know another, I'm all ears.

File: tools/build-toolchain.sh [hunk 14]
```diff
@@ -41,32 +78,61 @@ download () {
...
+echo "Compiling binutils-$BINUTILS_V"
 cd "binutils-$BINUTILS_V"
 ./configure \
   --prefix="$INSTALL_PATH" \
   --target=mips64-elf \
   --with-cpu=mips64vr4300 \
   --disable-werror
 make -j "$JOBS"
-make install || sudo make install || su -c "make install"
+make install || sudo make install || su -c "make install" # Perhaps use `checkinstall` instead?!
```
Comment (rasky) [inline tools/build-toolchain.sh]: One thing to try is to use `make install-strip` instead of `make install`. It should generate much smaller binaries. It should be available for both binutils and GCC.

The dockerfile runs strip manually but I think this is better done here for all kind of usages of the built toolchain.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -41,6 +154,7 @@ jobs:
       - name: Build image
         if: ${{ steps.path_diff.outputs.changed == 1 }}
         uses: docker/build-push-action@v2
+        # TODO: surely this might be related to the fork, as they might have an updated one!:
```
Comment (rasky) [inline .github/workflows/ci.yml]: Yes you are right but there is an unfortunate bug. Docker only supports lower case names as image names, so if you use the GitHub placeholders for the repo name, you end up with "DragonMinded" here (with the uppercase) which doesn't work. 

I tried looked around for a workaround but could not find one, so for now this is how it is.

Please change the TODO to report this, so that we will track it

File: Dockerfile [hunk 16]
```diff
@@ -35,6 +34,7 @@ ENV PATH="${N64_INST}/bin:$PATH"
 
 COPY --from=0 ${N64_INST} ${N64_INST}
 RUN apt-get update && \
-    apt-get install -yq gcc g++ make libpng-dev git  && \
+    apt-get install -yq gcc g++ make libpng-dev git && \
+    apt-get install -yq ninja-build rsync zip && \
```
Comment (rasky) [inline Dockerfile]: Why are you adding these packages in the final docker image?

File: Dockerfile [hunk 17]
```diff
@@ -16,15 +16,14 @@ RUN ./build-toolchain.sh
...
-RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/install-tools/fixincl
-RUN strip ${N64_INST}/libexec/gcc/mips64-elf/10.2.0/lto1
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/plugin/gengtype
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/liblto_plugin.so
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/lto-wrapper
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/collect2
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/cc1plus
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/cc1
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/install-tools/fixincl
+RUN strip ${N64_INST}/libexec/gcc/mips64-elf/11.2.0/lto1
```
Comment (rasky) [inline Dockerfile]: I think you should be able to remove these if you switch to `install-strip`

**Comment by rasky** [comment]
I'm testing this:
https://github.com/rasky/libdragon/commits/toolchain_tweaks1

It adds install-strip as suggested and tries to fix the Docker repo name. So I'll submit this when it passes.

File: Dockerfile (hunk 16 above)
Comment (networkfusion) [reply Dockerfile]: This should enable compiling ROM's using CMAKE (yet to test) and also starts work on being able to use as a `dev container` within VSCode. Although this should (probably) be a seperate PR... I will split this out when I get everything else working, as this PR is now a bit of a mess... But at least shows progress in the grand scheme of things!

File: tools/build-toolchain.sh (hunk 6 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: Would it be a good idea to make them global, since it makes them contollable for all targets?

File: Dockerfile (hunk 17 above)
Comment (networkfusion) [reply Dockerfile]: Done.

File: tools/build-toolchain.sh (hunk 6 above)
Comment (rasky) [reply tools/build-toolchain.sh]: No I don't think we want to do this. Basically anything goes with them, we just need it for targets that don't ship them with the package manager. I don't think we want to force a specific version in general.

File: tools/build-toolchain.sh [hunk 18]
```diff
@@ -1,34 +1,70 @@
...
+fi
+
+  # These "should" be the same as linux, but may be out of sync (as need to ensure working natively first).
+  # Binutils fails with 2.37 on canadian cross
+  BINUTILS_V=2.36.1 # so we are stuck with 2.36.1 for the moment
+  
+  # GCC 11.x fails on canadian cross
+  # see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100017
+  # see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80196
+  GCC_V=10.3.0 # so we are stuck with the 10.x branch for the moment.
```
Comment (rasky) [inline tools/build-toolchain.sh]: Can't you just download the patch found in Bugzilla and apply it using "patch" command?
https://gcc.gnu.org/bugzilla/attachment.cgi?id=51747

So that we build the same GCC for all targets

**Comment by networkfusion** [comment]
Finally! a valid reason for the dockerfile build to fail: 
image

File: tools/build-toolchain.sh (hunk 18 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: I (personally) think that just makes things more complicated... What are the benefits of GCC 11.2 over 10.3 from a libdragon perspective... I like using the latest toolchain, but would rather wait for the next one than implement lots of workarounds...

Comment (rasky) [reply tools/build-toolchain.sh]: Consistency. I don't want to tell people that different toolchain on different OSs will generate different ROMs. So maybe I compile my game on Linux, then you compile the same game on Windows and yours doesn't work because of a GCC bug. Or doesn't compile because I'm using a command line option not available in your version. 

Every libdragon version should be compile with the single toolchain specified in the build script. If the build scripts creates different toolchains on different hosts, we can create problems. 

I'd rather make this script more complicated by applying a patch rather than pushing the problem of different toolchains to users.

File: n64.mk [hunk 19]
```diff
@@ -1,3 +1,15 @@
...
+  FIXPATH =$(subst /,\,$1)
+  MKDIR = mkdir
+  MV = move
+  RM = del /F /Q # force deletion quietly
+else # We are running on UNIX
+  FIXPATH = $1
+  MKDIR = mkdir -p # make parent directories as needed
+  MV = mv -f # ignore nonexistent files, never prompt
+  RM = rm -f # ignore nonexistent files, never prompt
+endif
```
Comment (rasky) [inline n64.mk:11]: I understand the effort here, but we have a maintenance problem. This means that whatever we do to n64.mk from now on, we need to have an implementation also for the very limited CMD shell in Windows. This concerns me.
Do you have any alternative solution in mind? Why don't you like the msys solution?

File: tools/build-toolchain.sh [hunk 20]
```diff
@@ -1,34 +1,74 @@
...
+  FOREIGN_INSTALL_PATH="$THIS_SCRIPT_PATH/binaries"
+
+  # This will require the extra flags (under certain libs)
+  BUILD= # Probably never required but can use: --build=x86_64-linux-gnu
+  HOST="--host=$1" 
+
+  # Dependency source libs (Versions)
+  # This will have to build Make and download FP libs
+  GMP_V=6.2.0
+  MPC_V=1.2.1
```
Comment (rasky) [inline tools/build-toolchain.sh]: You should change this with a check for BUILD != HOST

File: tools/build-toolchain.sh [hunk 21]
```diff
@@ -38,35 +78,96 @@ command_exists () {
 
 # Download the file URL using wget or curl (depending on which is installed)
 download () {
-  if   command_exists wget ; then wget -c  "$1"
+  if   command_exists wget ; then wget --no-check-certificate -c  "$1"
```
Comment (rasky) [inline tools/build-toolchain.sh]: Why add `--no-check-certificate`?

File: tools/build-toolchain.sh [hunk 22]
```diff
@@ -38,35 +78,96 @@ command_exists () {
...
-test -d "newlib-$NEWLIB_V"     || tar -xzf "newlib-$NEWLIB_V.tar.gz"
+if [ "$BUILD" != "$HOST" ]; then
+  echo "Stage: Patch step"
+  
+  if [[ "$GCC_V" = "11.2.0" || "$GCC_V" = "11.1.0" ]]; then
+    # GCC 11.x fails on canadian cross
+    # see: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100017
+    # see: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80196
+    echo "Apply patch for GCC using SED:"
+    sed -z 's/RAW_CXX_FOR_TARGET="$CXX_FOR_TARGET"/RAW_CXX_FOR_TARGET="$CXX_FOR_TARGET -nostdinc++"/' ./"gcc-$GCC_V"/configure
```
Comment (rasky) [inline tools/build-toolchain.sh:134]: In general, I think it would be easier to track these as `.patch` files, committed in the source tree, and applied with `patch`.

File: tools/build-toolchain.sh (hunk 21 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: It turns out that wget needs extra CA dependencies (i.e. it actually checks the root CA). Try building the docker image against WGET with no CURL installed... This was an easy workaround.

File: tools/build-toolchain.sh [hunk 23]
```diff
@@ -1,34 +1,74 @@
...
+# TODO: does not reach bash strict as $1 could be null!
+if [ "$1" == "x86_64-w64-mingw32" ]; then # Windows cross compile (host) flag is specified as a parameter.
+  # This (may) also require the following (extra) package dependencies:
+  # sudo apt-get install -yq mingw-w64 libgmp-dev bison libz-mingw-w64-dev autoconf
+
+  echo "Cross compiling for different host"
+  # Use the current-directory/binaries for the install path, as this is not for linux!
+  rm -rf binaries && mkdir binaries # always ensure the folder is clean (if rebuilding)
+  THIS_SCRIPT_PATH="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+  FOREIGN_INSTALL_PATH="$THIS_SCRIPT_PATH/binaries"
```
Comment (rasky) [inline tools/build-toolchain.sh]: I'm a little confused by this FOREIGN_INSTALL_PATH. Can you explain why we need this? The comment says "as this is not for linux", but I don't understand why the Canadian cross build should need this special directory.

File: tools/build-toolchain.sh [hunk 24]
```diff
@@ -1,34 +1,74 @@
...
-# Set PATH for newlib to compile using GCC for MIPS N64 (pass 1)
-export PATH="$PATH:$INSTALL_PATH/bin"
+# Ensure you set 'N64_INST' before calling the script to change the default installation directory path
+  # by default it will presume 'usr/local/n64_toolchain'
+  INSTALL_PATH="${N64_INST:-/usr/local/n64_toolchain}"
+
+# Check for cross compile script flag
+# TODO: detect the flag `--host`
+# TODO: does not reach bash strict as $1 could be null!
+if [ "$1" == "x86_64-w64-mingw32" ]; then # Windows cross compile (host) flag is specified as a parameter.
```
Comment (rasky) [inline tools/build-toolchain.sh]: I think I prefer that this is not a command line option. If you define HOST/BUILD/TARGET at the global script level like this:

```bash
BUILD=$(BUILD:-x86_64-linux-gnu}
HOST=${HOST:-x86_64-linux-gnu}
TARGET=${TARGET:-mips64-elf}
```

then you can simply override `HOST` externally, such as `HOST=x86_64-w64-mingw32 build-toolchain.sh`

Comment (networkfusion) [reply tools/build-toolchain.sh]: Done.

File: tools/build-toolchain.sh (hunk 21 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: I have added a comment in source about this.

File: tools/build-toolchain.sh (hunk 20 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: I have gone further and now handle as an optional parameter.

File: tools/build-toolchain.sh (hunk 23 above)
Comment (networkfusion) [reply tools/build-toolchain.sh]: I have improved this in the script to make it more explicit. But, it really comes down to the fact that we might be building for 2 different "HOST"s in the same script. The install path for the "native" system should indeed be installed to a defined path, but if we are doing a foreign install, should we be adding extra things there... Although on further thought, it is not doing anything different to what I have done previously so might revisit once the current build completes.

**Comment by rasky** [comment]
Thanks, submitted as https://github.com/DragonMinded/libdragon/pull/306
